multiplex.go

```diff
@@ -430,33 +450,70 @@ func (mp *Multiplex) handleIncoming() {
 				// We're not accepting data on this stream, for
 				// some reason. It's likely that we reset it, or
 				// simply canceled reads (e.g., called Close).
-				mp.putBufferInbound(b)
+				if err := mp.skipNextMsg(mlen); err != nil {
+					mp.shutdownErr = err
+					return
+				}
 				continue
 			}
 
-			recvTimeout.Reset(ReceiveTimeout)
-			select {
-			case msch.dataIn <- b:
-			case <-msch.readCancel:
-				// the user has canceled reading. walk away.
-				mp.putBufferInbound(b)
-			case <-recvTimeout.C:
-				mp.putBufferInbound(b)
-				log.Warnf("timed out receiving message into stream queue.")
-				// Do not do this asynchronously. Otherwise, we
-				// could drop a message, then receive a message,
-				// then reset.
-				msch.Reset()
-				continue
-			case <-mp.shutdown:
-				mp.putBufferInbound(b)
-				return
-			}
-			if !recvTimeout.Stop() {
-				<-recvTimeout.C
+		read:
+			for rd := 0; rd < mlen; {
+				nextChunk := mlen - rd
+				if nextChunk > BufferSize {
+					nextChunk = BufferSize
+				}
+
+				b, err := mp.readNextMsg(nextChunk)
```

Comment:
nit: rename to `readNextChunk` to make it clear that this reads partial messages.

Reply:
done.